--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2230,7 +_,7 @@
     @Override
     public void handleTeleportToEntityPacket(ServerboundTeleportToEntityPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.isSpectator()) {
+        if (this.player.isSpectator() && this.player.getBukkitEntity().hasPermission("spectatetp")) { // BasicLandMC - Add permission check for teleporting in spectator mode
             for (ServerLevel serverLevel : this.server.getAllLevels()) {
                 Entity entity = packet.getEntity(serverLevel);
                 if (entity != null) {
