From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: root <pirkl.dan28@gmail.com>
Date: Fri, 14 Mar 2025 11:42:31 +0100
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 81e92d1053efd15c079e318a4ae087948bc07866..29191f75b85aa7b258829c3df28d190c3301d084 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -33,6 +33,13 @@ public interface ServerBuildInfo {
     Key BRAND_DIVINEMC_ID = Key.key("bxteam", "divinemc");
     // DivineMC end
 
+    // BL start
+    /**
+     * The brand id for BasicLandMC.
+     */
+    Key BRAND_BASICLANDMC_ID = Key.key("basicland", "basiclandmc");
+    // BL end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index a76439e59eefa4b6dbd0e100d72c21055d0ca008..a2e3b1b94aaeb5fa71835b7680ec45f27971ed66 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -259,7 +259,7 @@ public class VersionCommand extends BukkitCommand {
         // Purpur start
         int distance = getVersionFetcher().distance();
         final Component message = Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()),
-            ChatColor.parseMM("<grey>Current DivineMC Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()), // DivineMC - Rebrand
+            ChatColor.parseMM("<grey>Current BasicLandMC Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()), // BL - Rebrand
             // Purpur end
             msg
         );
