--- a/src/main/java/org/bxteam/divinemc/server/ServerLevelTickExecutorThreadFactory.java
+++ b/src/main/java/org/bxteam/divinemc/server/ServerLevelTickExecutorThreadFactory.java
@@ -9,12 +_,12 @@
 
     public ServerLevelTickExecutorThreadFactory(String worldName) {
         this.worldName = worldName;
-        ThreadDumperRegistry.REGISTRY.add(worldName + "ServerLevel Tick Worker");
+        ThreadDumperRegistry.REGISTRY.add(worldName + " - ServerLevel Tick Worker");
     }
 
     @Override
     public Thread newThread(Runnable runnable) {
-        TickThread.ServerLevelTickThread tickThread = new TickThread.ServerLevelTickThread(runnable, this.worldName + "ServerLevel Tick Worker");
+        TickThread.ServerLevelTickThread tickThread = new TickThread.ServerLevelTickThread(runnable, this.worldName + " - ServerLevel Tick Worker");
 
         if (tickThread.isDaemon()) {
             tickThread.setDaemon(false);
