From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 28 Jul 2026 15:37:53 +0200
Subject: [PATCH] misc-changes


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index cbce12c25237ea1c81b92496caf61f089d9c226b..e6c89ca769db5818da0ce70dd965ad7ad883b5f6 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -673,6 +673,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         ((org.bukkit.craftbukkit.help.SimpleHelpMap) this.server.getHelpMap()).initializeCommands();
         this.server.getPluginManager().callEvent(new org.bukkit.event.server.ServerLoadEvent(org.bukkit.event.server.ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
+        // BasiclandMC - start - metrics
+        if (server.getPluginManager().isPluginEnabled("UnifiedMetrics") && Boolean.parseBoolean(System.getProperty("BL.metrics", "true"))) {
+            cz.basicland.metrics.Metrics.init(LOGGER);
+        }
+        // BasiclandMC - end - metrics
     }
     // Paper end - rework world loading process
 
diff --git a/net/minecraft/server/commands/ForceLoadCommand.java b/net/minecraft/server/commands/ForceLoadCommand.java
index 280fe9d389862098747006cf6f2c2f3292c8ad92..9c795f2af2c16b08d7ab734fd5c2dd9c63730e28 100644
--- a/net/minecraft/server/commands/ForceLoadCommand.java
+++ b/net/minecraft/server/commands/ForceLoadCommand.java
@@ -148,9 +148,10 @@ public class ForceLoadCommand {
             int maxChunkX = SectionPos.blockToSectionCoord(maxX);
             int maxChunkZ = SectionPos.blockToSectionCoord(maxZ);
             long chunkCount = (maxChunkX - minChunkX + 1L) * (maxChunkZ - minChunkZ + 1L);
-            if (chunkCount > 256L) {
-                throw ERROR_TOO_MANY_CHUNKS.create(256, chunkCount);
-            }
+            // BasiclnadMC misc changes - up the limit - start
+            if (chunkCount > 2048L) {
+                throw ERROR_TOO_MANY_CHUNKS.create(2048, chunkCount);
+            } // BasiclandMC - end
 
             ServerLevel level = source.getLevel();
             ResourceKey<Level> dimension = level.dimension();
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4e8a85205d5d2e067142d1b1b01b490377f0cdae..df5704f86573194794d99e8a350b0d39c909b544 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2735,7 +2735,8 @@ public class ServerGamePacketListenerImpl
         // CraftBukkit end - replaced with thread safe throttle
             && !org.bxteam.divinemc.config.DivineConfig.NetworkCategory.disableDisconnectSpam // DivineMC - Option to disable disconnect.spam
             && !this.server.getPlayerList().isOp(this.player.nameAndId())
-            && !this.server.isSingleplayerOwner(this.player.nameAndId())) {
+            && !this.server.isSingleplayerOwner(this.player.nameAndId())
+            && !this.player.getBukkitEntity().hasPermission("chatspam")) { // BasicLandMC - Add permission to bypass spam protection
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause & add proper async disconnect
         }
     }
