From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 25 Nov 2025 20:37:02 +0100
Subject: [PATCH] back-port-rct-update


diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 71b0b72c0c65fada33bdf732aa8b366a6ab8e472..d7ee3b80f540c68c4e317d262d93ca06fc04bdde 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -157,45 +157,35 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     // Paper end - rewrite chunk system
     // Paper start - chunk tick iteration optimisations
     private final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom shuffleRandom = new ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom(0L);
-    private void iterateTickingChunksFaster(final CompletableFuture<Void> spawns) { // DivineMC - Regionized Chunk Ticking
+    protected void iterateTickingChunksFaster(final CompletableFuture<Void> spawns) { // DivineMC - private -> protected Regionized Chunk Ticking
         final ServerLevel world = this.level;
         final int randomTickSpeed = world.getGameRules().getInt(GameRules.RULE_RANDOMTICKING);
 
         // TODO check on update: impl of forEachBlockTickingChunk will only iterate ENTITY ticking chunks!
         // TODO check on update: consumer just runs tickChunk
-        final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> entityTickingChunks = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getEntityTickingChunks(); // DivineMC - Regionized Chunk Ticking
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.world.level.chunk.LevelChunk> entityTickingChunks = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getEntityTickingChunks();
 
         // note: we can use the backing array here because:
         // 1. we do not care about new additions
         // 2. _removes_ are impossible at this stage in the tick
-        final LevelChunk[] raw = entityTickingChunks.toArray(new LevelChunk[0]); // DivineMC - use toArray instead of getRawDataUnchecked this way is safe and doesn't have performance impact
+        final LevelChunk[] raw = entityTickingChunks.getRawDataUnchecked();
         final int size = entityTickingChunks.size();
 
-        // DivineMC start - Regionized Chunk Ticking
-        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableRegionizedChunkTicking) {
-            if (this instanceof org.bxteam.divinemc.async.rct.RegionizedChunkTicking rct) {
-                rct.execute(spawns, raw);
-            }
-        } else {
-            java.util.Objects.checkFromToIndex(0, size, raw.length);
-            for (int i = 0; i < size; ++i) {
-                world.tickChunk(raw[i], randomTickSpeed);
-
-                // call mid-tick tasks for chunk system
-                if ((i & 7) == 0) {
-                    // DivineMC start - Parallel world ticking
-                    if (!org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) {
-                        ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer) this.level.getServer()).moonrise$executeMidTickTasks();
-                        continue;
-                    }
-                    // DivineMC end - Parallel world ticking
-                }
-            }
+        java.util.Objects.checkFromToIndex(0, size, raw.length);
+        for (int i = 0; i < size; ++i) {
+            world.tickChunk(raw[i], randomTickSpeed);
 
-            if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.asyncNaturalSpawn) {
-                spawns.join();
+            // call mid-tick tasks for chunk system
+            if ((i & 7) == 0) {
+                ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks();
+                continue;
             }
         }
+
+        // DivineMC start - Regionized Chunk Ticking
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.asyncNaturalSpawn) {
+            spawns.join();
+        }
         // DivineMC end - Regionized Chunk Ticking
     }
     // Paper end - chunk tick iteration optimisations
@@ -645,7 +635,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             spawns = CompletableFuture.runAsync(() -> naturalSpawn(filteredSpawningCategories, timeInhabited), org.bxteam.divinemc.async.rct.RegionizedChunkTicking.REGION_EXECUTOR);
         } else {
             naturalSpawn(filteredSpawningCategories, timeInhabited);
-            spawns = new CompletableFuture<>();
+            spawns = CompletableFuture.completedFuture(null);
         }
         // DivineMC end - Regionized Chunk Ticking
 
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index bae2cf89e9bc71a360fe471f7d3e703705488f3a..46c776df0cd8a6b28efa6f5aaced98a8ac4068ac 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -393,7 +393,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     private int sectionX = Integer.MIN_VALUE;
     private int sectionY = Integer.MIN_VALUE;
     private int sectionZ = Integer.MIN_VALUE;
-    private boolean updatingSectionStatus;
+    private final java.util.concurrent.atomic.AtomicBoolean updatingSectionStatus = new java.util.concurrent.atomic.AtomicBoolean(false);
 
     @Override
     public final boolean moonrise$isHardColliding() {
@@ -452,12 +452,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     @Override
     public final boolean moonrise$isUpdatingSectionStatus() {
-        return this.updatingSectionStatus;
+        return this.updatingSectionStatus.get();
     }
 
     @Override
     public final void moonrise$setUpdatingSectionStatus(final boolean to) {
-        this.updatingSectionStatus = to;
+        this.updatingSectionStatus.set(to);
     }
 
     @Override
@@ -5028,7 +5028,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     public final void setPosRaw(double x, double y, double z, boolean forceBoundingBoxUpdate) {
         // Paper start - rewrite chunk system
-        if (this.updatingSectionStatus) {
+        if (this.updatingSectionStatus.get()) {
             LOGGER.error(
                 "Refusing to update position for entity " + this + " to position " + new Vec3(x, y, z)
                     + " since it is processing a section status update", new Throwable()
diff --git a/net/minecraft/world/level/entity/EntityTickList.java b/net/minecraft/world/level/entity/EntityTickList.java
index cc47fbb5b95370aadcd43ea296e46a0aedd925f7..42cb7885a0073d872676e32cf6b1253a7166a561 100644
--- a/net/minecraft/world/level/entity/EntityTickList.java
+++ b/net/minecraft/world/level/entity/EntityTickList.java
@@ -32,7 +32,7 @@ public class EntityTickList {
         if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(entity, "Asynchronous entity ticklist addition"); // DivineMC - Parallel world ticking
         this.ensureActiveIsNotIterated();
         // DivineMC start - Parallel world ticking
-        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.RCTorPWT) {
             synchronized (this.entities) {
                 this.entities.add(entity);
             }
@@ -46,7 +46,7 @@ public class EntityTickList {
         if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(entity, "Asynchronous entity ticklist removal"); // DivineMC - Parallel world ticking
         this.ensureActiveIsNotIterated();
         // DivineMC start - Parallel world ticking
-        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.RCTorPWT) {
             synchronized (this.entities) {
                 if (this.entities.contains(entity)) {
                     this.entities.remove(entity);
@@ -60,7 +60,7 @@ public class EntityTickList {
 
     public boolean contains(Entity entity) {
         // DivineMC start - Parallel world ticking
-        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableParallelWorldTicking) {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.RCTorPWT) {
             synchronized (this.entities) {
                 return this.entities.contains(entity);
             }
