From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Sun, 30 Nov 2025 17:30:05 +0100
Subject: [PATCH] debug-test


diff --git a/net/minecraft/server/network/ServerConnectionListener.java b/net/minecraft/server/network/ServerConnectionListener.java
index bd07e6a5aa1883786d789ea71711a0c0c0a95c26..bb3a46dd5519dca7e51323ef8e26b1b8250f6447 100644
--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -128,6 +128,7 @@ public class ServerConnectionListener {
 
                                     if (!disableFlushConsolidation) channel.pipeline().addFirst(new io.netty.handler.flush.FlushConsolidationHandler()); // Paper - Optimize network
                                     ChannelPipeline channelPipeline = channel.pipeline().addLast("timeout", new ReadTimeoutHandler(30));
+                                    channel.pipeline().addAfter("timeout", "debug_history", new cz.basicland.DebugPacketHistoryHandler(64));
                                     if (ServerConnectionListener.this.server.repliesToStatus()) {
                                         channelPipeline.addLast("legacy_query", new LegacyQueryHandler(ServerConnectionListener.this.getServer()));
                                     }
