From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Mon, 13 Jul 2026 21:30:49 +0200
Subject: [PATCH] optimize-defaults-for-divinemc-config


diff --git a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
index 421939fcfc6facb70a051335e41da232c20d142f..8fba413be147e62eeb2d55d8e47d0c316201e944 100644
--- a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
+++ b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
@@ -195,8 +195,8 @@ public class DivineConfig {
 
         // Parallel world ticking settings
         @Experimental("Parallel World Ticking")
-        public static boolean enableParallelWorldTicking = false;
-        public static int parallelThreadCount = 4;
+        public static boolean enableParallelWorldTicking = true; // BasiclandMC optimize defaults
+        public static int parallelThreadCount = 6; // BasiclandMC optimize defaults
         public static boolean logContainerCreationStacktraces = false;
         public static boolean disableHardThrow = false;
         public static boolean usePerWorldTpsBar = true;
@@ -204,8 +204,8 @@ public class DivineConfig {
 
         // Regionized chunk ticking
         @Experimental("Regionized Chunk Ticking")
-        public static boolean enableRegionizedChunkTicking = false;
-        public static int regionizedChunkTickingExecutorThreadCount = 4;
+        public static boolean enableRegionizedChunkTicking = true; // BasiclandMC optimize defaults
+        public static int regionizedChunkTickingExecutorThreadCount = 6; // BasiclandMC optimize defaults
         public static int regionizedChunkTickingExecutorThreadPriority = Thread.NORM_PRIORITY + 2;
 
         // Async pathfinding settings
@@ -439,7 +439,7 @@ public class DivineConfig {
         public static int hopperThrottleSkipTicks = 0;
 
         // DAB settings
-        public static boolean dabEnabled = false;
+        public static boolean dabEnabled = true; // BasiclandMC optimize defaults
         public static int dabStartDistance = 12;
         public static int dabStartDistanceSquared;
         public static int dabMaximumActivationFrequency = 20;
@@ -777,7 +777,7 @@ public class DivineConfig {
         public static int playerProfileResultCachingTimeout = 1440;
 
         // No chat reports
-        public static boolean noChatReportsEnabled = false;
+        public static boolean noChatReportsEnabled = true; // BasiclandMC optimize defaults
         public static boolean noChatReportsAddQueryData = true;
         public static boolean noChatReportsConvertToGameMessage = true;
         public static boolean noChatReportsDebugLog = false;
@@ -785,12 +785,12 @@ public class DivineConfig {
         public static String noChatReportsDisconnectDemandOnClientMessage = "You do not have No Chat Reports, and this server is configured to require it on client!";
 
         // Protocols
-        public static boolean protocolsAppleSkinEnabled = false;
+        public static boolean protocolsAppleSkinEnabled = true; // BasiclandMC optimize defaults
         public static int protocolsAppleSkinSyncTickInterval = 20;
-        public static boolean protocolsJadeEnabled = false;
+        public static boolean protocolsJadeEnabled = true; // BasiclandMC optimize defaults
         public static boolean protocolsMapsXaeroMapEnabled = false;
         public static int protocolsMapsXaeroMapServerId = new Random().nextInt();
-        public static boolean protocolsSyncMaticaEnabled = false;
+        public static boolean protocolsSyncMaticaEnabled = true; // BasiclandMC optimize defaults
         public static boolean protocolsSyncMaticaQuota = false;
         public static int protocolsSyncMaticaQuotaLimit = 40000000;
 
