From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 28 Jul 2026 15:42:07 +0200
Subject: [PATCH] optimize-defaults-for-divinemc-config


diff --git a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
index 817b907cf3fa5e7ef28137775f247c6aa419f60a..87dec61557c82f3afd0914629bacdb21dac9f8d4 100644
--- a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
+++ b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
@@ -240,7 +240,7 @@ public class DivineConfig {
 
         // Parallel world ticking settings
         @Experimental("Parallel World Ticking")
-        public static boolean enableParallelWorldTicking = false;
+        public static boolean enableParallelWorldTicking = true; // BasiclandMC optimize defaults
         public static int parallelThreadCount = 4;
         public static boolean logContainerCreationStacktraces = false;
         public static boolean disableHardThrow = false;
@@ -249,24 +249,24 @@ public class DivineConfig {
 
         // Regionized chunk ticking
         @Experimental("Regionized Chunk Ticking")
-        public static boolean enableRegionizedChunkTicking = false;
-        public static int regionizedChunkTickingExecutorThreadCount = 4;
+        public static boolean enableRegionizedChunkTicking = true; // BasiclandMC optimize defaults
+        public static int regionizedChunkTickingExecutorThreadCount = 6; // BasiclandMC optimize defaults
         public static int regionizedChunkTickingExecutorThreadPriority = Thread.NORM_PRIORITY + 2;
 
         // Async pathfinding settings
         public static boolean asyncPathfinding = true;
-        public static int asyncPathfindingMaxThreads = 1;
+        public static int asyncPathfindingMaxThreads = 2; // BasiclandMC optimize defaults
         public static int asyncPathfindingKeepalive = 60;
         public static int asyncPathfindingQueueSize = 0;
         public static PathfindTaskRejectPolicy asyncPathfindingRejectPolicy = PathfindTaskRejectPolicy.CALLER_RUNS;
 
         // Parallel entity tracker settings
         public static boolean parallelTrackerEnabled = true;
-        public static int parallelTrackerThreads = 0;
+        public static int parallelTrackerThreads = 2; // BasiclandMC optimize defaults
         public static int parallelTrackerKeepalive = 60;
 
         // Async chunk sending settings
-        public static boolean asyncChunkSendingEnabled = false;
+        public static boolean asyncChunkSendingEnabled = true; // BasiclandMC optimize defaults
         public static int asyncChunkSendingMaxThreads = 1;
 
         // Async mob spawning settings
@@ -278,8 +278,8 @@ public class DivineConfig {
 
         // Parallel sensor phase settings
         @Experimental("Parallel Sensor Phase")
-        public static boolean parallelSensorsEnabled = false;
-        public static int parallelSensorsMaxThreads = 0;
+        public static boolean parallelSensorsEnabled = true; // BasiclandMC optimize defaults
+        public static int parallelSensorsMaxThreads = 2; // BasiclandMC optimize defaults
 
         public static void load() {
             enableAutoAllocation();
@@ -535,7 +535,7 @@ public class DivineConfig {
         public static int hopperThrottleSkipTicks = 0;
 
         // DAB settings
-        public static boolean dabEnabled = false;
+        public static boolean dabEnabled = true; // BasiclandMC optimize defaults
         public static int dabStartDistance = 12;
         public static int dabStartDistanceSquared;
         public static int dabMaximumActivationFrequency = 20;
@@ -891,7 +891,7 @@ public class DivineConfig {
         public static int playerProfileResultCachingTimeout = 1440;
 
         // No chat reports
-        public static boolean noChatReportsEnabled = false;
+        public static boolean noChatReportsEnabled = true; // BasiclandMC optimize defaults
         public static boolean noChatReportsAddQueryData = true;
         public static boolean noChatReportsConvertToGameMessage = true;
         public static boolean noChatReportsDebugLog = false;
@@ -900,12 +900,12 @@ public class DivineConfig {
 
         // Protocols
         public static boolean protocolsStrictMode = false;
-        public static boolean protocolsAppleSkinEnabled = false;
+        public static boolean protocolsAppleSkinEnabled = true; // BasiclandMC optimize defaults
         public static int protocolsAppleSkinSyncTickInterval = 20;
-        public static boolean protocolsJadeEnabled = false;
-        public static boolean protocolsMapsXaeroMapEnabled = false;
+        public static boolean protocolsJadeEnabled = true; // BasiclandMC optimize defaults
+        public static boolean protocolsMapsXaeroMapEnabled = true; // BasiclandMC optimize defaults
         public static int protocolsMapsXaeroMapServerId = new Random().nextInt();
-        public static boolean protocolsSyncMaticaEnabled = false;
+        public static boolean protocolsSyncMaticaEnabled = true; // BasiclandMC optimize defaults
         public static boolean protocolsSyncMaticaQuota = false;
         public static int protocolsSyncMaticaQuotaLimit = 40000000;
 
