From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Sat, 14 Mar 2026 12:55:03 +0100
Subject: [PATCH] optimize-defaults-for-divinemc-config


diff --git a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
index b1b004c219e4629c5c9a47e685a390fb76455439..464031429b28f9d08efefe7616d6bf5895a98465 100644
--- a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
+++ b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
@@ -193,8 +193,8 @@ public class DivineConfig {
     public static class AsyncCategory {
         // Parallel world ticking settings
         @Experimental("Parallel World Ticking")
-        public static boolean enableParallelWorldTicking = false;
-        public static int parallelThreadCount = 4;
+        public static boolean enableParallelWorldTicking = true; // BasiclandMC optimize defaults
+        public static int parallelThreadCount = 6; // BasiclandMC optimize defaults
         public static boolean logContainerCreationStacktraces = false;
         public static boolean disableHardThrow = false;
         public static boolean usePerWorldTpsBar = true;
@@ -202,8 +202,8 @@ public class DivineConfig {
 
         // Regionized chunk ticking
         @Experimental("Regionized Chunk Ticking")
-        public static boolean enableRegionizedChunkTicking = false;
-        public static int regionizedChunkTickingExecutorThreadCount = 4;
+        public static boolean enableRegionizedChunkTicking = true; // BasiclandMC optimize defaults
+        public static int regionizedChunkTickingExecutorThreadCount = 6; // BasiclandMC optimize defaults
         public static int regionizedChunkTickingExecutorThreadPriority = Thread.NORM_PRIORITY + 2;
 
         // Async pathfinding settings
@@ -263,10 +263,12 @@ public class DivineConfig {
             regionizedChunkTickingExecutorThreadPriority = getInt(ConfigCategory.ASYNC.key("regionized-chunk-ticking.executor-thread-priority"), regionizedChunkTickingExecutorThreadPriority,
                 "Configures the thread priority of the executor");
 
-            if (regionizedChunkTickingExecutorThreadCount < 1 || regionizedChunkTickingExecutorThreadCount > 10) {
-                LOGGER.warn("Invalid regionized chunk ticking thread count: {}, resetting to default (4)", regionizedChunkTickingExecutorThreadCount);
-                regionizedChunkTickingExecutorThreadCount = 4;
+            // BasiclandMC - optimize config start
+            if (regionizedChunkTickingExecutorThreadCount < 1 || regionizedChunkTickingExecutorThreadCount > (Runtime.getRuntime().availableProcessors() / 2)) {
+                LOGGER.warn("Invalid regionized chunk ticking thread count: {}, resetting to default (6)", regionizedChunkTickingExecutorThreadCount);
+                regionizedChunkTickingExecutorThreadCount = 6;
             }
+            // BasiclandMC - end
         }
 
         private static void asyncPathfinding() {
@@ -384,7 +386,7 @@ public class DivineConfig {
         public static int hopperThrottleSkipTicks = 0;
 
         // DAB settings
-        public static boolean dabEnabled = false;
+        public static boolean dabEnabled = true; // BasiclandMC optimize defaults
         public static int dabStartDistance = 12;
         public static int dabStartDistanceSquared;
         public static int dabMaximumActivationFrequency = 20;
@@ -398,14 +400,16 @@ public class DivineConfig {
             "goat"
         ));
 
+        // BasiclandMC optimize defaults start
         // Virtual threads
-        public static boolean virtualThreadsEnabled = false;
-        public static boolean virtualBukkitScheduler = false;
-        public static boolean virtualChatScheduler = false;
-        public static boolean virtualTabCompleteScheduler = false;
-        public static boolean virtualAsyncExecutor = false;
-        public static boolean virtualCommandBuilderScheduler = false;
-        public static boolean virtualServerTextFilterPool = false;
+        public static boolean virtualThreadsEnabled = true;
+        public static boolean virtualBukkitScheduler = true;
+        public static boolean virtualChatScheduler = true;
+        public static boolean virtualTabCompleteScheduler = true;
+        public static boolean virtualAsyncExecutor = true;
+        public static boolean virtualCommandBuilderScheduler = true;
+        public static boolean virtualServerTextFilterPool = true;
+        // BasiclandMC optimize defaults end
 
         public static void load() {
             chunkSettings();
@@ -722,7 +726,7 @@ public class DivineConfig {
         public static int playerProfileResultCachingTimeout = 1440;
 
         // No chat reports
-        public static boolean noChatReportsEnabled = false;
+        public static boolean noChatReportsEnabled = true; // BasiclandMC optimize defaults
         public static boolean noChatReportsAddQueryData = true;
         public static boolean noChatReportsConvertToGameMessage = true;
         public static boolean noChatReportsDebugLog = false;
@@ -730,12 +734,12 @@ public class DivineConfig {
         public static String noChatReportsDisconnectDemandOnClientMessage = "You do not have No Chat Reports, and this server is configured to require it on client!";
 
         // Protocols
-        public static boolean protocolsAppleSkinEnabled = false;
+        public static boolean protocolsAppleSkinEnabled = true; // BasiclandMC optimize defaults
         public static int protocolsAppleSkinSyncTickInterval = 20;
-        public static boolean protocolsJadeEnabled = false;
-        public static boolean protocolsMapsXaeroMapEnabled = false;
+        public static boolean protocolsJadeEnabled = true; // BasiclandMC optimize defaults
+        public static boolean protocolsMapsXaeroMapEnabled = true; // BasiclandMC optimize defaults
         public static int protocolsMapsXaeroMapServerId = new Random().nextInt();
-        public static boolean protocolsSyncMaticaEnabled = false;
+        public static boolean protocolsSyncMaticaEnabled = true; // BasiclandMC optimize defaults
         public static boolean protocolsSyncMaticaQuota = false;
         public static int protocolsSyncMaticaQuotaLimit = 40000000;
 
