From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <pirkldan28@gmail.com>
Date: Sat, 12 Apr 2025 14:03:39 +0200
Subject: [PATCH] sync-fix


diff --git a/src/main/java/org/bxteam/divinemc/util/BlockEntityTickersList.java b/src/main/java/org/bxteam/divinemc/util/BlockEntityTickersList.java
index 328f27d788fd7873742c6335d5d5ddb73ce33d26..877cbe97adafe5e74cd4c67e29fd522ec0d5be2a 100644
--- a/src/main/java/org/bxteam/divinemc/util/BlockEntityTickersList.java
+++ b/src/main/java/org/bxteam/divinemc/util/BlockEntityTickersList.java
@@ -34,6 +34,13 @@ public final class BlockEntityTickersList extends ObjectArrayList<TickingBlockEn
         super(c);
     }
 
+    @Override
+    public boolean addAll(int index, final Collection<? extends TickingBlockEntity> c) {
+        synchronized (c) {
+            return super.addAll(index, c);
+        }
+    }
+
     /**
      * Marks an entry as removed
      *
