From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Sun, 22 Jun 2025 20:22:43 +0200
Subject: [PATCH] tps-bar-show-tps-of-server-option


diff --git a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
index 592c5891844ac87b6d6aa498db39ba5ae9567a60..be89f78b5f4af6d472207fca5db1e3bb9e64d14c 100644
--- a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
+++ b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
@@ -194,6 +194,7 @@ public class DivineConfig {
         public static boolean disableHardThrow = false;
         public static boolean pwtCompatabilityMode = false;
         public static boolean usePerWorldTpsBar = true;
+        public static boolean showTPSOfServerInsteadOfWorld = true;
 
         // Regionized chunk ticking
         @Experimental("Regionized Chunk Ticking")
@@ -249,6 +250,10 @@ public class DivineConfig {
                 "Enables compatibility mode for plugins that are not compatible with Parallel World Ticking. This makes all async tasks run synchronously.");
             usePerWorldTpsBar = getBoolean(ConfigCategory.ASYNC.key("parallel-world-ticking.use-per-world-tps-bar"), usePerWorldTpsBar,
                 "Enables per-world TPS bar, which shows the TPS of the world the player is currently in. TPS bar can be turned on/off with /tpsbar command.");
+            showTPSOfServerInsteadOfWorld = getBoolean(ConfigCategory.ASYNC.key("parallel-world-ticking.show-tps-of-server-instead-of-world"), showTPSOfServerInsteadOfWorld,
+                "Enables showing the TPS of the server instead of the world in the TPS bar.",
+                "This is useful if you want to see the overall performance of the server instead of the performance of a specific world.",
+                "While keeping mspt of the world in the TPS bar.");
         }
 
         private static void regionizedChunkTicking() {
