From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 29 Jul 2025 10:36:20 +0200
Subject: [PATCH] optimize-defaults-for-divinemc-config


diff --git a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
index be89f78b5f4af6d472207fca5db1e3bb9e64d14c..3b2f71eaee4f18ad84a47610c866898208d125bc 100644
--- a/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
+++ b/src/main/java/org/bxteam/divinemc/config/DivineConfig.java
@@ -188,8 +188,8 @@ public class DivineConfig {
     public static class AsyncCategory {
         // Parallel world ticking settings
         @Experimental("Parallel World Ticking")
-        public static boolean enableParallelWorldTicking = false;
-        public static int parallelThreadCount = 4;
+        public static boolean enableParallelWorldTicking = true; // BasiclandMC optimize defaults
+        public static int parallelThreadCount = 6; // BasiclandMC optimize defaults
         public static boolean logContainerCreationStacktraces = false;
         public static boolean disableHardThrow = false;
         public static boolean pwtCompatabilityMode = false;
@@ -198,13 +198,13 @@ public class DivineConfig {
 
         // Regionized chunk ticking
         @Experimental("Regionized Chunk Ticking")
-        public static boolean enableRegionizedChunkTicking = false;
-        public static int regionizedChunkTickingExecutorThreadCount = 4;
+        public static boolean enableRegionizedChunkTicking = true; // BasiclandMC optimize defaults
+        public static int regionizedChunkTickingExecutorThreadCount = 6; // BasiclandMC optimize defaults
         public static int regionizedChunkTickingExecutorThreadPriority = Thread.NORM_PRIORITY + 2;
 
         // Async pathfinding settings
         public static boolean asyncPathfinding = true;
-        public static int asyncPathfindingMaxThreads = 2;
+        public static int asyncPathfindingMaxThreads = 1; // BasiclandMC optimize defaults
         public static int asyncPathfindingKeepalive = 60;
         public static int asyncPathfindingQueueSize = 0;
         public static PathfindTaskRejectPolicy asyncPathfindingRejectPolicy = PathfindTaskRejectPolicy.FLUSH_ALL;
@@ -393,15 +393,16 @@ public class DivineConfig {
             "goat"
         ));
 
-        // Virtual threads
-        public static boolean virtualThreadsEnabled = false;
-        public static boolean virtualBukkitScheduler = false;
-        public static boolean virtualChatScheduler = false;
-        public static boolean virtualTabCompleteScheduler = false;
-        public static boolean virtualAsyncExecutor = false;
-        public static boolean virtualCommandBuilderScheduler = false;
-        public static boolean virtualProfileLookupPool = false;
-        public static boolean virtualServerTextFilterPool = false;
+        // Virtual threads // BasiclandMC optimize defaults start
+        public static boolean virtualThreadsEnabled = true;
+        public static boolean virtualBukkitScheduler = true;
+        public static boolean virtualChatScheduler = true;
+        public static boolean virtualTabCompleteScheduler = true;
+        public static boolean virtualAsyncExecutor = true;
+        public static boolean virtualCommandBuilderScheduler = true;
+        public static boolean virtualProfileLookupPool = true;
+        public static boolean virtualServerTextFilterPool = true;
+        // BasiclandMC optimize defaults end
 
         public static void load() {
             chunkSettings();
@@ -784,7 +785,7 @@ public class DivineConfig {
         public static int playerProfileResultCachingTimeout = 1440;
 
         // No chat reports
-        public static boolean noChatReportsEnabled = false;
+        public static boolean noChatReportsEnabled = true; // BasiclandMC optimize defaults
         public static boolean noChatReportsAddQueryData = true;
         public static boolean noChatReportsConvertToGameMessage = true;
         public static boolean noChatReportsDebugLog = false;
@@ -795,9 +796,9 @@ public class DivineConfig {
         public static boolean protocolsAppleSkinEnabled = false;
         public static int protocolsAppleSkinSyncTickInterval = 20;
         public static boolean protocolsJadeEnabled = false;
-        public static boolean protocolsMapsXaeroMapEnabled = false;
+        public static boolean protocolsMapsXaeroMapEnabled = true;  // BasiclandMC optimize defaults
         public static int protocolsMapsXaeroMapServerId = new Random().nextInt();
-        public static boolean protocolsSyncMaticaEnabled = false;
+        public static boolean protocolsSyncMaticaEnabled = true;  // BasiclandMC optimize defaults
         public static boolean protocolsSyncMaticaQuota = false;
         public static int protocolsSyncMaticaQuotaLimit = 40000000;
 
