From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 11 Nov 2025 18:20:07 +0100
Subject: [PATCH] shutdown-fix


diff --git a/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java b/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
index ce8a353008e64d82c1e17f95ae805cb772d1ebd5..207891345bb6a29d842e0fb2d12b12725ea316f0 100644
--- a/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
+++ b/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
@@ -38,7 +38,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (AsyncChunkSend.POOL != null) {
+        if (DivineConfig.AsyncCategory.asyncChunkSendingEnabled && AsyncChunkSend.POOL != null) {
             LOGGER.info("Shutting down async chunk send executor...");
             AsyncChunkSend.POOL.shutdown();
 
@@ -47,7 +47,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (MultithreadedTracker.TRACKER_EXECUTOR != null) {
+        if (DivineConfig.AsyncCategory.multithreadedEnabled && MultithreadedTracker.TRACKER_EXECUTOR != null) {
             LOGGER.info("Shutting down mob tracker executor...");
             MultithreadedTracker.TRACKER_EXECUTOR.shutdown();
 
@@ -56,7 +56,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (AsyncPathProcessor.PATH_PROCESSING_EXECUTOR != null) {
+        if (DivineConfig.AsyncCategory.asyncPathfinding && AsyncPathProcessor.PATH_PROCESSING_EXECUTOR != null) {
             LOGGER.info("Shutting down mob pathfinding processing executor...");
             AsyncPathProcessor.PATH_PROCESSING_EXECUTOR.shutdown();
 
@@ -65,7 +65,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (AsyncJoinHandler.JOIN_EXECUTOR != null) {
+        if (DivineConfig.AsyncCategory.asyncJoinEnabled && AsyncJoinHandler.JOIN_EXECUTOR != null) {
             LOGGER.info("Shutting down async join executor...");
             AsyncJoinHandler.JOIN_EXECUTOR.shutdown();
 
