From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Thu, 13 Nov 2025 17:18:24 +0100
Subject: [PATCH] shutdown-fix


diff --git a/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java b/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
index 9ff985f543e173b7b182590086e39c63a77ca1f6..ad8d9e21e646497432de5cc8f737f0bf01511817 100644
--- a/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
+++ b/src/main/java/org/bxteam/divinemc/async/ExecutorShutdown.java
@@ -38,7 +38,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (AsyncChunkSend.POOL != null) {
+        if (DivineConfig.AsyncCategory.asyncChunkSendingEnabled && AsyncChunkSend.POOL != null) {
             LOGGER.info("Shutting down async chunk send executor...");
             AsyncChunkSend.POOL.shutdown();
 
@@ -47,7 +47,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (MultithreadedTracker.TRACKER_EXECUTOR != null) {
+        if (DivineConfig.AsyncCategory.multithreadedEnabled && MultithreadedTracker.TRACKER_EXECUTOR != null) {
             LOGGER.info("Shutting down mob tracker executor...");
             MultithreadedTracker.TRACKER_EXECUTOR.shutdown();
 
@@ -56,7 +56,7 @@ public class ExecutorShutdown {
             } catch (InterruptedException ignored) { }
         }
 
-        if (AsyncPathProcessor.PATH_PROCESSING_EXECUTOR != null) {
+        if (DivineConfig.AsyncCategory.asyncPathfinding && AsyncPathProcessor.PATH_PROCESSING_EXECUTOR != null) {
             LOGGER.info("Shutting down mob pathfinding processing executor...");
             AsyncPathProcessor.PATH_PROCESSING_EXECUTOR.shutdown();
 
