From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 28 Jul 2026 15:30:54 +0200
Subject: [PATCH] rebrand


diff --git a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
index c1580dc6c44bbd0d14f68baecba0ec2ce3f4aaad..966aa9115dcb2062cc8e21e57aa77df1e859e571 100644
--- a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
+++ b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
@@ -21,7 +21,7 @@ public final class PaperConsole extends SimpleTerminalConsole {
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
         builder
-                .appName("DivineMC") // DivineMC - Rebrand
+                .appName("BasiclandMC") // BasicLandMC - Rebrand
                 .variable(LineReader.HISTORY_FILE, historyFile(this.server)) // DivineMC - make configurable location of files
                 .completer(new ConsoleCommandCompleter(this.server))
                 .option(LineReader.Option.COMPLETE_IN_WORD, true);
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index c0c294d5e4e28d944787a572ec16cfdb8623fc75..c92c6dbb68de390f7417ff1f7d9378c82a324568 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -35,6 +35,7 @@ public record ServerBuildInfoImpl(
     private static final String BRAND_PAPER_NAME = "Paper";
     private static final String BRAND_PURPUR_NAME = "Purpur"; // Purpur - Rebrand
     private static final String BRAND_DIVINEMC_NAME = "DivineMC"; // DivineMC - Rebrand
+    private static final String BRAND_BASICLANDMC_NAME = "BasicLandMC"; // BasicLandMC - Rebrand
 
     private static final String BUILD_DEV = "DEV";
 
@@ -46,9 +47,9 @@ public record ServerBuildInfoImpl(
         this(
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                 .map(Key::key)
-                .orElse(BRAND_DIVINEMC_ID), // DivineMC - Rebrand
+                .orElse(BRAND_BASICLANDMC_ID), // BasicLandMC - Rebrand
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
-                .orElse(BRAND_DIVINEMC_NAME), // DivineMC - Rebrand
+                .orElse(BRAND_BASICLANDMC_NAME), // BasicLandMC - Rebrand
             SharedConstants.getCurrentVersion().id(),
             SharedConstants.getCurrentVersion().name(),
             getManifestAttribute(manifest, ATTRIBUTE_BUILD_NUMBER)
