From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Sat, 14 Mar 2026 18:03:11 +0100
Subject: [PATCH] task-logging


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a176c09d389600aeb3aa10eeb09d15c9ad80dedf..2583ef7be8c5b1ab9f1981278a4a21fa95808023 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -602,7 +602,9 @@ public final class CraftServer implements Server {
 
         for (Plugin plugin : plugins) {
             if ((!plugin.isEnabled()) && (plugin.getDescription().getLoad() == type)) {
+                long time = System.currentTimeMillis(); // BasicLandMC - Add time measurement
                 this.enablePlugin(plugin);
+                logger.info("Enabled Plugin '" + plugin.getDescription().getFullName() + "' in " + (System.currentTimeMillis() - time) + " ms"); // BasicLandMC - Add time measurement
             }
         }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index a3e4065916f2aa2971f4e19f0702c7a052af2090..fc8b7e24c0af5c2821a20b1e624d7fc2f3f42701 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -45,6 +45,10 @@ import org.bukkit.scheduler.BukkitWorker;
  * </ul>
  */
 public class CraftScheduler implements BukkitScheduler {
+    private static final it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<Plugin> PLUGIN_TIME_PER_TICK = new it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<>();
+    static {
+        PLUGIN_TIME_PER_TICK.defaultReturnValue(0);
+    }
 
     /**
      * The start ID for the counter.
@@ -452,6 +456,7 @@ public class CraftScheduler implements BukkitScheduler {
      */
     public void mainThreadHeartbeat() {
         this.currentTick++;
+        PLUGIN_TIME_PER_TICK.clear();
         // Paper start
         if (!this.isAsyncScheduler) {
             this.asyncScheduler.mainThreadHeartbeat();
@@ -469,6 +474,7 @@ public class CraftScheduler implements BukkitScheduler {
                 continue;
             }
             if (task.isSync()) {
+                long nanos = System.nanoTime(); //BasiclandMC - Print tasks that take too long to process
                 this.currentTask = task;
                 try {
                     task.run();
@@ -487,6 +493,16 @@ public class CraftScheduler implements BukkitScheduler {
                     // Paper end
                 } finally {
                     this.currentTask = null;
+                    //BasiclandMC - Print tasks that take too long to process - start
+                    long time = System.nanoTime() - nanos;
+                    PLUGIN_TIME_PER_TICK.addTo(task.getOwner(), (int) time);
+                    if (time >= 10_000_000L) {
+                        task.getOwner().getLogger().log(
+                            Level.WARNING,
+                            String.format("Task #%s for %s took %dms to process (class name: %s)", task.getTaskId(), task.getOwner().getDescription().getFullName(), time / 1_000_000L, task.getTaskClass().getName())
+                        );
+                    }
+                    //BasiclandMC - Print tasks that take too long to process - end
                 }
                 this.parsePending();
             } else {
@@ -505,6 +521,16 @@ public class CraftScheduler implements BukkitScheduler {
         }
         this.pending.addAll(temp);
         temp.clear();
+        //BasiclandMC - Print plugins that take too long to process tasks in the last tick - start
+        PLUGIN_TIME_PER_TICK.object2IntEntrySet().forEach(entry -> {
+            if (entry.getIntValue() >= 10_000_000L) {
+                entry.getKey().getLogger().log(
+                    Level.WARNING,
+                    String.format("Plugin %s took %dms to process all sync tasks", entry.getKey().getName(), entry.getIntValue() / 1_000_000L)
+                );
+            }
+        });
+        //BasiclandMC - Print plugins that take too long to process tasks in the last tick - end
         //this.debugHead = this.debugHead.getNextHead(this.currentTick); // Paper
     }
 
