From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: root <pirkl.dan28@gmail.com>
Date: Fri, 14 Mar 2025 11:54:25 +0100
Subject: [PATCH] Implement-Loading-Plugins-From-External-Folder


diff --git a/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java b/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
index 70413fddd23ca1165cb5090cce4fddcb1bbca93f..029dfaa7edd422042beb37d0892d570c725d1786 100644
--- a/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
+++ b/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
@@ -112,6 +112,30 @@ public class PluginInitializerManager {
         // Register the default plugin directory
         io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(io.papermc.paper.plugin.provider.source.DirectoryProviderSource.INSTANCE, pluginSystem.pluginDirectoryPath());
 
+        //BL start
+        @SuppressWarnings("unchecked")
+        java.util.List<Path> pluginList = ((java.util.List<File>) optionSet.valuesOf("add-plugin-dir")).stream()
+            .filter(java.util.Objects::nonNull)
+            .map(File::listFiles)
+            .filter(java.util.Objects::nonNull)
+            .flatMap(java.util.Arrays::stream)
+            .filter(File::isFile)
+            .map(File::toPath)
+            .toList();
+
+        io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(io.papermc.paper.plugin.provider.source.PluginFlagProviderSource.INSTANCE, pluginList);
+
+        boolean override = (boolean) optionSet.valueOf("override");
+        if (override) {
+            File[] overrides = new File("override").listFiles((file, name) -> name.endsWith(".jar"));
+            if (overrides != null) {
+                java.util.List<Path> overrideList = java.util.Arrays.stream(overrides).filter(java.util.Objects::nonNull).map(File::toPath).toList();
+                io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(io.papermc.paper.plugin.provider.source.PluginFlagProviderSource.INSTANCE, overrideList);
+            }
+        }
+
+        //BL end
+
         // Register plugins from the flag
         @SuppressWarnings("unchecked")
         java.util.List<Path> files = ((java.util.List<File>) optionSet.valuesOf("add-plugin")).stream().map(File::toPath).toList();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index f367efa0afae9792f87a842dfe7ba098a3ce7f2d..0a22b6f79d3d2843217b9b85570a6375e62ddc6e 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -199,6 +199,20 @@ public class Main {
                         .defaultsTo("Unknown Server")
                         .describedAs("Name");
                 // Paper end
+
+                // BL start
+                acceptsAll(asList("add-plugin-dir", "add-extra-plugin-dir"), "Specify paths to directories containing extra plugin jars to be loaded in addition to those in the plugins folder. This argument can be specified multiple times, once for each extra plugin directory path.")
+                    .withRequiredArg()
+                    .ofType(File.class)
+                    .defaultsTo(new File("extra"))
+                    .describedAs("Directory");
+
+                acceptsAll(asList("override"), "Enables override directory for plugin loading")
+                    .withRequiredArg()
+                    .ofType(Boolean.class)
+                    .defaultsTo(false)
+                    .describedAs("boolean");
+                // BL end
             }
         };
 
