From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Wed, 11 Feb 2026 22:44:48 +0100
Subject: [PATCH] task-logging


diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index a839dbbb72f48b8f8736d9f4693c528686570732..8faf7861c77e996e96f569117d3149c61cf18fc0 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -45,7 +45,10 @@ import org.bukkit.scheduler.BukkitWorker;
  * </ul>
  */
 public class CraftScheduler implements BukkitScheduler {
-
+    private static final it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<Plugin> PLUGIN_TIME_PER_TICK = new it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<>();
+    static {
+        PLUGIN_TIME_PER_TICK.defaultReturnValue(0);
+    }
     /**
      * The start ID for the counter.
      */
@@ -452,6 +455,7 @@ public class CraftScheduler implements BukkitScheduler {
      */
     public void mainThreadHeartbeat() {
         this.currentTick++;
+        PLUGIN_TIME_PER_TICK.clear();
         // Paper start
         if (!this.isAsyncScheduler) {
             this.asyncScheduler.mainThreadHeartbeat();
@@ -469,6 +473,7 @@ public class CraftScheduler implements BukkitScheduler {
                 continue;
             }
             if (task.isSync()) {
+                long nanos = System.nanoTime(); //BasiclandMC - Print tasks that take too long to process
                 this.currentTask = task;
                 try {
                     task.run();
@@ -487,6 +492,16 @@ public class CraftScheduler implements BukkitScheduler {
                     // Paper end
                 } finally {
                     this.currentTask = null;
+                    //BasiclandMC - Print tasks that take too long to process - start
+                    long time = System.nanoTime() - nanos;
+                    PLUGIN_TIME_PER_TICK.addTo(task.getOwner(), (int) time);
+                    if (time >= 10_000_000L) {
+                        task.getOwner().getLogger().log(
+                            Level.WARNING,
+                            String.format("Task #%s for %s took %dms to process (class name: %s)", task.getTaskId(), task.getOwner().getDescription().getFullName(), time / 1_000_000L, task.getTaskClass().getName())
+                        );
+                    }
+                    //BasiclandMC - Print tasks that take too long to process - end
                 }
                 this.parsePending();
             } else {
@@ -505,6 +520,20 @@ public class CraftScheduler implements BukkitScheduler {
         }
         this.pending.addAll(temp);
         temp.clear();
+        //BasiclandMC - Print plugins that take too long to process tasks in the last tick - start
+        PLUGIN_TIME_PER_TICK.object2IntEntrySet().forEach(entry -> {
+            if (entry.getIntValue() >= 10_000_000L) {
+                String msg = String.format("Plugin %s took %dms to process all sync tasks", entry.getKey().getName(), entry.getIntValue() / 1_000_000L);
+                if (entry.getKey().getName().equals("ProtocolLib")) {
+                    System.out.println(msg);
+                }
+                entry.getKey().getLogger().log(
+                    Level.WARNING,
+                    msg
+                );
+            }
+        });
+        //BasiclandMC - Print plugins that take too long to process tasks in the last tick - end
         //this.debugHead = this.debugHead.getNextHead(this.currentTick); // Paper
     }
 
