From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <pirkldan28@gmail.com>
Date: Fri, 18 Apr 2025 20:53:00 +0200
Subject: [PATCH] update-impl-of-virtual-threads


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 3c7f66571ca346e9776ba6044bbb0f0263f34157..aed3a7966390d5651694b46bb65daabb2e5d072d 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -37,11 +37,18 @@ public final class MCUtil {
             run.run();
         }
     };
-    public static final ExecutorService ASYNC_EXECUTOR = org.bxteam.divinemc.DivineConfig.virtualThreadsEnabled && org.bxteam.divinemc.DivineConfig.virtualAsyncExecutor ? Executors.newVirtualThreadPerTaskExecutor() : Executors.newFixedThreadPool(2, new ThreadFactoryBuilder() // DivineMC - Virtual Threads
-        .setNameFormat("Paper Async Task Handler Thread - %1$d")
-        .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(MinecraftServer.LOGGER))
-        .build()
-    );
+    public static final ExecutorService ASYNC_EXECUTOR;
+
+    static {
+        ThreadFactoryBuilder threadFactoryBuilder = new ThreadFactoryBuilder() // DivineMC - Virtual Threads
+            .setNameFormat("Paper Async Task Handler Thread - %1$d")
+            .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(MinecraftServer.LOGGER));
+        if (org.bxteam.divinemc.DivineConfig.virtualThreadsEnabled && org.bxteam.divinemc.DivineConfig.virtualAsyncExecutor) {
+            threadFactoryBuilder.setThreadFactory(Thread.ofVirtual().factory());
+        }
+
+        ASYNC_EXECUTOR = Executors.newFixedThreadPool(2, threadFactoryBuilder.build());
+    }
 
     private MCUtil() {
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
index bfa84901c5e0e50ab3f713293035aefbcf4488e6..5b7d1ede1f6f53c2f1c8f605780ff19a39f5e15b 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
@@ -39,12 +39,23 @@ import java.util.concurrent.TimeUnit;
 
 public class CraftAsyncScheduler extends CraftScheduler {
     // DivineMC start - Virtual Threads
-    private final ExecutorService executor = org.bxteam.divinemc.DivineConfig.virtualThreadsEnabled && org.bxteam.divinemc.DivineConfig.virtualBukkitScheduler
-        ? Executors.newVirtualThreadPerTaskExecutor()
-        : new ThreadPoolExecutor(
-    // DivineMC end - Virtual Threads
-            4, Integer.MAX_VALUE,30L, TimeUnit.SECONDS, new SynchronousQueue<>(),
-            new ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d").build());
+    private final ExecutorService executor;
+
+    {
+        ThreadFactoryBuilder threadFactoryBuilder = new ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d");
+        if (org.bxteam.divinemc.DivineConfig.virtualThreadsEnabled && org.bxteam.divinemc.DivineConfig.virtualBukkitScheduler) {
+            threadFactoryBuilder.setThreadFactory(Thread.ofVirtual().factory());
+        }
+
+        executor = new ThreadPoolExecutor(
+            4,
+            Integer.MAX_VALUE,
+            30L,
+            TimeUnit.SECONDS,
+            new SynchronousQueue<>(),
+            threadFactoryBuilder.build()); // DivineMC end - Virtual Threads
+    }
+
     private final Executor management = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
             .setNameFormat("Craft Async Scheduler Management Thread").build());
     private final List<CraftTask> temp = new ArrayList<>();
