--- a/divinemc-api/build.gradle.kts
+++ b/divinemc-api/build.gradle.kts
@@ -40,6 +_,7 @@
 }
 
 dependencies {
+    api(project(":api")) //ASP
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:33.4.0-jre") // DivineMC - Bump dependencies
     api("com.google.code.gson:gson:2.12.1") // DivineMC - Bump dependencies
@@ -107,20 +_,24 @@
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
             srcDir(file("../purpur-api/src/main/java")) // DivineMC - Rebrand
+            srcDir(file("../divinemc-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
             srcDir(file("../purpur-api/src/main/resources")) // DivineMC - Rebrand
+            srcDir(file("../divinemc-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
             srcDir(file("../purpur-api/src/test/java")) // DivineMC - Rebrand
+            srcDir(file("../divinemc-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
             srcDir(file("../purpur-api/src/test/resources")) // DivineMC - Rebrand
+            srcDir(file("../divinemc-api/src/test/resources"))
         }
     }
 }
