--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -477,7 +_,19 @@
         for (int i = 0; i < listTag.size(); i++) {
             CompoundTag compound = listTag.getCompound(i);
             int i1 = compound.getByte("Slot") & 255;
-            ItemStack itemStack = ItemStack.parse(this.player.registryAccess(), compound).orElse(ItemStack.EMPTY);
+
+            // BL start - log broken items
+            com.mojang.serialization.DataResult<ItemStack> itemStackDataResult = ItemStack.parseGet(this.player.registryAccess(), compound);
+            java.util.Optional<ItemStack> result = itemStackDataResult.result();
+            ItemStack itemStack;
+            if (result.isEmpty()) {
+                com.mojang.logging.LogUtils.getLogger().error("Player '{}' has broken item in slot '{}'", player.getStringUUID(), i1);
+                itemStack = itemStackDataResult.resultOrPartial().orElse(ItemStack.EMPTY);
+            } else {
+                itemStack = result.get();
+            }
+            // BL end - log broken items
+
             if (i1 >= 0 && i1 < this.items.size()) {
                 this.items.set(i1, itemStack);
             } else if (i1 >= 100 && i1 < this.armor.size() + 100) {
