--- a/src/main/java/org/spigotmc/RestartCommand.java
+++ b/src/main/java/org/spigotmc/RestartCommand.java
@@ -11,7 +_,7 @@
 public class RestartCommand extends Command {
 
     public RestartCommand(String name) {
-        super(name);
+        super(name, "", "/" + name, java.util.List.of("reset"));
         this.description = "Restarts the server";
         this.usageMessage = "/restart";
         this.setPermission("bukkit.command.restart");
