From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Wed, 25 Sep 2024 22:57:25 +0200
Subject: [PATCH] Build-changes


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 6811f38a2eaf614c8237190800a48d101c39cd99..410ef2e801b2c8d46f89ac11c588cbc22263338e 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -40,6 +40,13 @@ public interface ServerBuildInfo {
     Key BRAND_DIVINEMC_ID = Key.key("divinemc", "divinemc");
     // DivineMC end
 
+    // BL start
+    /**
+     * The brand id for BasicLandMC.
+     */
+    Key BRAND_BASICLANDMC_ID = Key.key("basiclandmc", "basiclandmc");
+    // BL end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
