From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Tue, 29 Oct 2024 13:21:17 +0100
Subject: [PATCH] log-plugin-onEnable-load-time


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 57a4d95df6e6044eef01201b494417789090f010..2c09f1ea23a588cdc23e6c13d3181b60e713f605 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -605,7 +605,9 @@ public final class CraftServer implements Server {
 
         for (Plugin plugin : plugins) {
             if ((!plugin.isEnabled()) && (plugin.getDescription().getLoad() == type)) {
+                long time = System.currentTimeMillis(); // BL - Add time measurement
                 this.enablePlugin(plugin);
+                logger.info("Enabled Plugin '" + plugin.getDescription().getFullName() + "' in " + (System.currentTimeMillis() - time) + " ms"); // BL - Add time measurement
             }
         }
 
