--- a/divinemc-server/build.gradle.kts
+++ b/divinemc-server/build.gradle.kts
@@ -52,10 +_,38 @@
             outputDir = rootDirectory.dir("purpur-server")
         }
     }
-
-    activeFork = divinemc
     // DivineMC end - Rebrand
 
+    // BasicLandMC start - Rebrand
+    val basiclandmc = forks.register("basiclandmc") {
+        forks = divinemc
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = divinemc.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("basiclandmc-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchRepo("purpurServer") {
+            upstreamRepo = divinemc.patchedRepo("purpurServer")
+            patchesDir = rootDirectory.dir("basiclandmc-server/purpur-patches")
+            outputDir = rootDirectory.dir("purpur-server")
+        }
+        upstream.patchDir("divinemcServer") {
+            upstreamPath = "divinemc-server"
+            excludes = setOf(
+                "src/minecraft",
+                "paper-patches",
+                "minecraft-patches",
+                "build.gradle.kts",
+                "build.gradle.kts.patch"
+            )
+            patchesDir = rootDirectory.dir("basiclandmc-server/divinemc-patches")
+            outputDir = rootDirectory.dir("divinemc-server")
+        }
+    }
+
+    activeFork = basiclandmc
+    // BasicLandMC end - Rebrand
+
     /*updatingMinecraft { // Purpur start - TODO - snapshot - we don't need this
         // oldPaperCommit = "d4fe85375af18bfa88f44d7c1e6a61904ae550cc"
     }*/ // Purpur end - TODO - snapshot - we don't need this
@@ -130,6 +_,10 @@
         java { srcDir("../purpur-server/src/main/java") }
         resources { srcDir("../purpur-server/src/main/resources") }
         // DivineMC end - Rebrand
+        // BasicLandMC start - Rebrand
+        java { srcDir("../divinemc-server/src/main/java") }
+        resources { srcDir("../divinemc-server/src/main/resources") }
+        // BasicLandMC end - Rebrand
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
@@ -138,6 +_,10 @@
         java { srcDir("../purpur-server/src/test/java") }
         resources { srcDir("../purpur-server/src/test/resources") }
         // DivineMC end - Rebrand
+        // BasicLandMC start - Rebrand
+        java { srcDir("../divinemc-server/src/test/java") }
+        resources { srcDir("../divinemc-server/src/test/resources") }
+        // BasicLandMC end - Rebrand
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -166,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":divinemc-api")) // DivineMC - Rebrand
+    implementation(project(":basiclandmc-api")) // DivineMC - Rebrand
 
     // DivineMC start - Dependencies
     implementation ("com.github.Carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
@@ -260,14 +_,14 @@
         val channel = rootProject.providers.gradleProperty("channel").get() // DivineMC - Channel
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "DivineMC", // DivineMC - Rebrand
+            "Implementation-Title" to "BasiclandMC", // BasiclandMC - Rebrand
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "DivineMC", // DivineMC - Rebrand
+            "Specification-Title" to "BasiclandMC", // BasiclandMC - Rebrand
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "BX Team", // DivineMC - Rebrand
-            "Brand-Id" to "bxteam:divinemc", // DivineMC - Rebrand
-            "Brand-Name" to "DivineMC", // DivineMC - Rebrand
+            "Specification-Vendor" to "BL Team", // BasiclandMC - Rebrand
+            "Brand-Id" to "basicland:basiclandmc", // BasiclandMC - Rebrand
+            "Brand-Name" to "BasiclandMC", // BasiclandMC - Rebrand
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
